File: 600/623 - Add One Row to Tree.c
static void helper(struct TreeNode **pp, int is_left, struct TreeNode *root, int v, int d) {
    struct TreeNode *p;
    
    if (! --d) {
        p = (struct TreeNode *)malloc(sizeof(struct TreeNode));
        p->val = v;
        
        if (is_left) {
            p->left = root;
            p->right = NULL;
        } else {
            p->left = NULL;
            p->right = root;
        }
        
        *pp = p;
    } else {
        if (root) {
            helper(&root->left, 1, root->left, v, d);
            helper(&root->right, 0, root->right, v, d);   
        }
    }
}

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode* addOneRow(struct TreeNode* root, int v, int d) {
    helper(&root, 1, root, v, d);
    return root;
}

File: csharp/623.add-one-row-to-tree.cs
/*
 * @lc app=leetcode id=623 lang=csharp
 *
 * [623] Add One Row to Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode AddOneRow(TreeNode root, int v, int d) {
        if (d == 1) {
            var ret = new TreeNode(v);
            ret.left = root;
            return ret;
        }
        if (root == null) {
            return null;
        }
        if (d == 2) {
            var left = new TreeNode(v);
            var right = new TreeNode(v);
            left.left = root.left;
            right.right = root.right;
            root.left = left;
            root.right = right;
        } else {
            AddOneRow(root.left, v, d - 1);
            AddOneRow(root.right, v, d - 1);
        }
        return root;
    }
}
// @lc code=end

